### ISSUE TITLE
single shared library for libkml

### ISSUE THREAD

**Issue by rashadkm** [libkml/libkml#215]
I am not sure about this change right now. I am wondering should we make a single library libkml.so instead of kmlbase, kmlengine etc..

thirdparties using libkml usually link with all. Also there is no particular option to disable a part of libkml during build. Having a single likbml.so seems a nice feature for me.

There might a very less chance where users need only kmlbase or kmlengine . Not for kmlengine depends on kmlbase and kmldom. 

Before going any further about this, I would like to have some feedback from community.

Any thoughts?

@sebastic

**Comment by sebastic** [comment]
I'd like to stay close to the design as it is to not diverge too much from google/libkml, a change like this will make it harder to merge back than it already is. We should push the merging effort some more before making big changes like this.

**Comment by rashadkm** [comment]
there won't be any merge conflict  with google/libkml. because all these are in the cmake files which does not exists in google version. If that is what you are saying. And during merge the autoconf build is removed.

But I agree the way libkml is used with third parties needs to be updated if this happens. linking with all the libkml\* needs to be changed to libkml. This can be taken as big change but a good one IMHO. 

Asking on google/libkml about any update on the pull request has no effect so far. https://github.com/google/libkml/pull/6

Maybe we could ask on gdal or some other list to get their "attention" for a day or two.

**Comment by rashadkm** [comment]
i did some improvement on this issue - https://github.com/rashadkm/libkml/tree/master/ (work in progress)
hopefully it will help in using libkml in external projects. 

I also added thirdparty source (only if not found) in this branch.  (another WIP)

**Comment by rashadkm** [comment]
should be able to make a release at end of this task.

**Comment by sebastic** [comment]
How does this impact compatibility with google/libkml?

As long as this fork uses the same name, it should remain compatible (i.e. no code nor buildsystem changes required for projects depending on libkml).

**Comment by rashadkm** [comment]
okay. that means either keep compatibilty or rename project ?
what if there can be still all those stuff and the single libkml is a configuration change?

normally splitting into different shared libs is a good thing. but in this case, it is not much useful. This is why I go to a single libfile like in expat, zlib, and others.

**Comment by sebastic** [comment]
Yes, when libkml/libkml diverges too much from google/libkml the project should be renamed to reflect it being an incompatible project. While they both have their origins in the libkml project on Google Code, they can then no longer be considered the same project.

Having a configuration option to keep compatibility with google/libkml is a good idea to not require downstream users to adapt to the libkml/libkml changes. Projects like GDAL, OTB & GoldenCheetah have so far used google/libkml and expect that code to be used (or something close to it, as this libkml/libkml fork initially was).

@schwehr, as the Googler who's involved in GDAL upstream, what are your thoughts about this issue and the general situation with google/libkml being a dormant project and the efforts of this libkml/libkml fork to keep the codebase alive for the benefit of projects like the above relying on it.

**Comment by rashadkm** [comment]
i am not much interested in renaming this fork. So I guess the change is not possible in the current situation. Unless the fork got merged into google/libkml nothing can be done.  renaming makes it completely different project (forked from google and other libraries using libkml has to adopt which is something I believe developers are willing to do.

BTW, current state with boost, and other c++ stuff is too much for a library to process libkml. See I am not against boost or any other libs. Just that the idea of a library that reads kml, kmz files requiring a "semi-standard" c++ library is too much.

**Comment by sebastic** [comment]
I think the osgEarth developers would agree with you, they've switched away from libkml for similar reasons IIRC.

It's a shame that it's almost a year since the last progress in https://github.com/google/libkml/issues/4, it seemed so promising at the time.

**Comment by rashadkm** [comment]
yes. google has zero interest in this code and what makes it horrible is their no-no repsonse strategy.

anyway, If I were to write a library to read kml/kmz files, I would start with C, without boost and uriparser. Having to know the user base (gdal, OTB etc..) , code could be written with easy and sane transition.

**Comment by rashadkm** [comment]
out of curiosity, what osgearth is using to read kml files?

**Comment by sebastic** [comment]
osgEarth is using their own implementation, see:
https://github.com/gwaldron/osgearth/tree/master/src/osgEarthDrivers/kml

**Comment by schwehr** [comment]
I'm sorry I haven't made any progress on libkml.  I've be digging hard into GDAL technical debt.  And I finally got a chance to do a first pass through the gdal's libkml driver.  e.g. look for goatbar (which is me) in https://trac.osgeo.org/gdal/log/trunk/gdal/ogr/ogrsf_frmts/libkml   Even has been doing an amazing job doing cleanup and I am about 2K commits into GDAL cleanup.  I've started put out some of the gunit based unittests for GDAL that we have.... https://github.com/schwehr/gdal-autotest2/  We've been fuzz testing things like libtiff, kakadu, etc.  But there is still a long way to go.

I can only speak for myself and not for Google, but I am trying hard to make the best use of the time I have.  I need to work on GDAL before I spend time on libkml.  I am interested in moving libkml forward. I'm for whatever helps the community the most.

With only a first glance at the issue, I don't like the many targets and would rather there be a single library.  For my own work on GDAL with a custom bazel based build, I break out libs for speed with testing, but then present only a single target for users of GDAL and things are much cleaner / easier to maintain.

**Comment by sebastic** [comment]
One significant benefit that libkml/libkml has is its use of pkg-config. If we can adapt the downstream projects like GDAL, to use pkg-config when available and fallback to the manual linking if not, compatibility with google/libkml should remain for those users. It allows libkml/libkml more flexibility to change like using a single library.

**Comment by rkanavath** [comment]
I can fix the libkml.pc to work with single libkml so or if they are split in multiple "components". That would be easier fix IMHO. But if there still requires a need for renaming project, we would prefer not to take that road. the library itself does not have much users and development around and renaming it make the situation worse for the entire community. 

@schwehr , First of all thanks for getting back to this topic. for large projects, it make sense to split into modules that users can opt out when building the whole library. But for libkml where there is no optional dependencies, I thought a single so would be better for maintaining and for users of libkml.

**Comment by sebastic** [comment]
The project should be renamed when it diverges too much from google/libkml, and it can no longer be considered the same project.

I place that boundary where downstream users need specific code changes to support libkml/libkml in addition to google/libkml. As long as the libkml API remains compatible, the projects can still be considered the same. With the pkg-config changes like GDAL the buildsystem can support both variants without needing further buildsystem changes.

Dropping the separate shared libraries, will break at least OTB and will require changes for GoldenCheetah.

For OTB `FindLibKML.cmake` needs to be patched to prefer pkg-config over the current code, see:
https://sources.debian.net/src/otb/5.6.1%2Bdfsg-1/CMake/FindLibKML.cmake/#L1

And for GoldenCheetah changing `gcconfig.pri` should be sufficient:
https://sources.debian.net/src/goldencheetah/3.3.0-3/debian/gcconfig.pri/#L101
https://sources.debian.net/src/goldencheetah/3.3.0-3/src/src.pro/#L118

**Comment by rashadkm** [comment]
@schwehr , is it possible to merge changes to google/libkml? what is to be expected in the current state?
if there is something missing on our side, i would be happy to fix that
